lib/services/upstart.c

```diff
@@ -61,55 +61,70 @@ upstart_cleanup(void)
     }
 }
 
-static gboolean
-upstart_job_by_name(const gchar * arg_name, gchar ** out_unit, int timeout)
+/*!
+ * \internal
+ * \brief Get the DBus object path corresponding to a job name
+ *
+ * \param[in]  arg_name  Name of job to get path for
+ * \param[out] path      If not NULL, where to store DBus object path
+ * \param[in]  timeout   Give up after this many seconds
+ *
+ * \return true if object path was found, false otherwise
+ * \note The caller is responsible for freeing *path if it is non-NULL.
+ */
+static bool
+object_path_for_job(const gchar *arg_name, char **path, int timeout)
 {
-/*
-  com.ubuntu.Upstart0_6.GetJobByName (in String name, out ObjectPath job)
-*/
+    /*
+        com.ubuntu.Upstart0_6.GetJobByName (in String name, out ObjectPath job)
+    */
     DBusError error;
     DBusMessage *msg;
     DBusMessage *reply = NULL;
     const char *method = "GetJobByName";
+    bool rc = false;
 
-    if(upstart_init() == FALSE) {
-        return FALSE;
+    if (path != NULL) {
+        *path = NULL;
+    }
+
+    if (!upstart_init()) {
+        return false;
     }
     msg = dbus_message_new_method_call(BUS_NAME, // target for the method call
                                        BUS_PATH, // object to call on
                                        UPSTART_06_API, // interface to call on
                                        method); // method name
 
     dbus_error_init(&error);
-    CRM_LOG_ASSERT(dbus_message_append_args(msg, DBUS_TYPE_STRING, &arg_name, DBUS_TYPE_INVALID));
+    CRM_LOG_ASSERT(dbus_message_append_args(msg, DBUS_TYPE_STRING, &arg_name,
+                                            DBUS_TYPE_INVALID));
     reply = pcmk_dbus_send_recv(msg, upstart_proxy, &error, timeout);
     dbus_message_unref(msg);
 
     if (dbus_error_is_set(&error)) {
-        crm_err("Could not issue %s for %s: %s", method, arg_name, error.message);
+        crm_err("Could not get DBus object path for %s: %s",
+                arg_name, error.message);
         dbus_error_free(&error);
 
-    } else if(!pcmk_dbus_type_check(reply, NULL, DBUS_TYPE_OBJECT_PATH, __func__, __LINE__)) {
-        crm_err("Invalid return type for %s", method);
+    } else if (!pcmk_dbus_type_check(reply, NULL, DBUS_TYPE_OBJECT_PATH,
+                                     __func__, __LINE__)) {
+        crm_err("Could not get DBus object path for %s: Invalid return type",
+                arg_name);
 
     } else {
-        if(out_unit) {
-            char *path = NULL;
-
-            dbus_message_get_args (reply, NULL,
-                                   DBUS_TYPE_OBJECT_PATH, &path,
-                                   DBUS_TYPE_INVALID);
-
-            *out_unit = strdup(path);
+        if (path != NULL) {
+            dbus_message_get_args(reply, NULL, DBUS_TYPE_OBJECT_PATH, path,
+                                  DBUS_TYPE_INVALID);
+            *path = strdup(*path);
```

Comment:
chances that this can somehow happen are probably low enough as not to reflect it in the return-value which would as well mean we'd have to get the arg even if !path.

Reply:
it's worth checking *path before doing strdup(). I don't think we need anything else since the caller can check whether path is null

Reply:
was thinking about the case where path isn't set so that the caller doesn't have the chance to check - but as said will probably not happen and should be enough to prevent an exception for the case it still does.